worker/deployer/unit_manifolds.go

```diff
@@ -0,0 +1,304 @@
+
+package deployer
+
+import (
+	"time"
+
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/utils/voyeur"
+	"github.com/juju/worker/v2/dependency"
+
+	coreagent "github.com/juju/juju/agent"
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/api/base"
+	commonapi "github.com/juju/juju/api/common"
+	msapi "github.com/juju/juju/api/meterstatus"
+	"github.com/juju/juju/cmd/jujud/agent/engine"
+	"github.com/juju/juju/core/machinelock"
+	"github.com/juju/juju/worker"
+	"github.com/juju/juju/worker/agent"
+	"github.com/juju/juju/worker/apiaddressupdater"
+	"github.com/juju/juju/worker/apicaller"
+	"github.com/juju/juju/worker/apiconfigwatcher"
+	"github.com/juju/juju/worker/fortress"
+	"github.com/juju/juju/worker/leadership"
+	loggerworker "github.com/juju/juju/worker/logger"
+	"github.com/juju/juju/worker/logsender"
+	"github.com/juju/juju/worker/meterstatus"
+	"github.com/juju/juju/worker/metrics/collect"
+	"github.com/juju/juju/worker/metrics/sender"
+	"github.com/juju/juju/worker/metrics/spool"
+	"github.com/juju/juju/worker/migrationflag"
+	"github.com/juju/juju/worker/migrationminion"
+	"github.com/juju/juju/worker/retrystrategy"
+	"github.com/juju/juju/worker/uniter"
+	"github.com/juju/juju/worker/upgrader"
+)
+
+// UnitManifoldsConfig allows specialisation of the result of Manifolds.
+type UnitManifoldsConfig struct {
+
+	// LoggingContext holds the unit writers so that the loggers
+	// for the unit get tagged with the right source.
+	LoggingContext *loggo.Context
+
+	// Agent contains the agent that will be wrapped and made available to
+	// its dependencies via a dependency.Engine.
+	Agent coreagent.Agent
+
+	// LogSource will be read from by the logsender component.
+	LogSource logsender.LogRecordCh
+
+	// LeadershipGuarantee controls the behaviour of the leadership tracker.
+	LeadershipGuarantee time.Duration
+
+	// AgentConfigChanged is set whenever the unit agent's config
+	// is updated.
+	AgentConfigChanged *voyeur.Value
+
+	// ValidateMigration is called by the migrationminion during the
+	// migration process to check that the agent will be ok when
+	// connected to the new target controller.
+	ValidateMigration func(base.APICaller) error
+
+	// UpdateLoggerConfig is a function that will save the specified
+	// config value as the logging config in the agent.conf file.
+	UpdateLoggerConfig func(string) error
+
+	// MachineLock is a central source for acquiring the machine lock.
+	// This is used by a number of workers to ensure serialisation of actions
+	// across the machine.
+	MachineLock machinelock.Lock
+
+	// Clock supplies timekeeping services to various workers.
+	Clock clock.Clock
+}
+
+// UnitManifolds returns a set of co-configured manifolds covering the various
+// responsibilities of nested unit agent.
+//
+// Thou Shalt Not Use String Literals In This Function. Or Else.
+func UnitManifolds(config UnitManifoldsConfig) dependency.Manifolds {
```

Comment:
Do we need this or can we reuse the existing unit agent manifolds?
I guess there will be subtle differences between what's needed for standalone vs embedded....

Reply:
We definitely need a different list. There are quite a few differences between them. Many of the unit agent manifolds have requirements around the upgrade worker having been run, but we don't even have that here. I felt it was better to have a clean implementation for the units here.

Originally I was planning to remove the old uniter code, but I have a feeling we may need it for the k8s side of things.